### PR TITLE
Duplo 16561 GCP Firestore Resource

### PR DESCRIPTION
**Pull request by Duplo-nikv**
## Overview

GCP Firestore Resource implementation
## Summary of changes
Added CURD support for GCP Firestore Resource `duplocloud_gcp_firestore`
This PR does the following:

- Implemented resource for gcp firestore 
- Implemented data-source for gcp firestore

## Testing performed

- [ ] Using unit tests
- [ ] Manually, on my local system
- [✓ ] Manually, on a remote test system

## Describe any breaking changes

- ...

___

## **Type**
enhancement, documentation

___

## **Description**
- Introduced GCP Firestore resource and data sources for managing GCP Firestore in DuploCloud.
- Added CRUD operations and data retrieval for GCP Firestore resources.
- Provided documentation and examples for GCP Firestore resource and data sources.
- Registered new resource and data sources in the Terraform provider.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>data_source_duplo_gcp_firestore.go</strong><dd><code>Add GCP Firestore Data Source</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/data_source_duplo_gcp_firestore.go

<li>Added a new data source for GCP Firestore with schema and read <br>function.<br> <li> Implemented validation for tenant_id and name fields.<br> <li> Added computed attributes for Firestore details.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-a9a26aecc27f6b04b1cddcc5bb286365b4f40ea836078a80c7e6df54b5064257">+110/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>data_source_duplo_gcp_firestores.go</strong><dd><code>New Data Source for Listing GCP Firestores</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/data_source_duplo_gcp_firestores.go

<li>Introduced a new data source to list GCP Firestores.<br> <li> Implemented Firestore listing schema with computed attributes.<br> <li> Added function to read and list Firestore instances.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-5374f2e01afe5c879a78ce54254531b6f778caacb8ee4683bf3228fae72df853">+143/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_gcp_firestore.go</strong><dd><code>Implement GCP Firestore Resource</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/resource_duplo_gcp_firestore.go

<li>Implemented a new resource for managing GCP Firestore.<br> <li> Added CRUD operations and schema for Firestore resource.<br> <li> Included validation for Firestore type and location_id.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-ac51446b036b780b9b71f8111357137134ecfb0e074f7b8548acb6b12f6f2503">+281/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>gcp_firestore.go</strong><dd><code>SDK Support for GCP Firestore Operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplosdk/gcp_firestore.go

<li>Added SDK functions for GCP Firestore CRUD operations.<br> <li> Implemented Firestore creation, deletion, update, and retrieval <br>functions.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-c3bc8de0680f48e8a36b4f94678c546c4c9428245cf8c4b80b5122a80a9dab98">+72/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>provider.go</strong><dd><code>Register GCP Firestore Resource and Data Sources</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/provider.go

<li>Registered <code>duplocloud_gcp_firestore</code> resource and data sources in the <br>provider.<br> <li> Added <code>duplocloud_gcp_firestores</code> data source to the provider.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-0959744faba70cfc67abd70733e39a6f668d41c9fae69a27968841ef8ed56eeb">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation
</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>import.sh</strong><dd><code>Example Script for Importing GCP Firestore</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
examples/resources/duplocloud_gcp_firestore/import.sh

<li>Provided an example script for importing an existing GCP Firestore.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-b3afa8080e8f1d5232f7ae25752f74e1f18a0601db3fcba506870c188882ae68">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update Changelog for GCP Firestore Addition</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
CHANGELOG.md

<li>Documented the addition of GCP Firestore resource and data sources.<br> <li> Noted enhancements to DynamoDB table logic.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>gcp_firestore.md</strong><dd><code>Document GCP Firestore Data Source</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
docs/data-sources/gcp_firestore.md

<li>Added documentation for the <code>duplocloud_gcp_firestore</code> data source.<br> <li> Detailed schema and example usage.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-3b3065bc79af539cb6033cc67b94f2caf77b307d8276028411945a75d0a989a6">+59/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>gcp_firestores.md</strong><dd><code>Document Data Source for Listing GCP Firestores</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
docs/data-sources/gcp_firestores.md

<li>Created documentation for the <code>duplocloud_gcp_firestores</code> data source.<br> <li> Explained schema and provided example usage.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-1c216c6ab95fca482058c3fe5bdf4bd4baa08ae781e4a9af386fa5e2b0a42ba8">+55/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>gcp_firestore.md</strong><dd><code>Document GCP Firestore Resource</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
docs/resources/gcp_firestore.md

<li>Introduced documentation for the <code>duplocloud_gcp_firestore</code> resource.<br> <li> Described resource attributes and provided example configuration.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-bfa19d6e2c059102cb6be810437a70e16d16de336b6814699e0a874c0a7e70b4">+82/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>data-source.tf</strong><dd><code>Terraform Example for GCP Firestore Data Source</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
examples/data-sources/duplocloud_gcp_firestore/data-source.tf

<li>Provided a Terraform example for using the <code>duplocloud_gcp_firestore</code> <br>data source.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-4bc8ff47b95f68062a3b1186938ad8ab08cc12faa54fb43c647ee6e35bd85de2">+20/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>data-source.tf</strong><dd><code>Terraform Example for Listing GCP Firestores Data Source</code>&nbsp; </dd></summary>
<hr>
      
examples/data-sources/duplocloud_gcp_firestores/data-source.tf

<li>Added a Terraform example for the <code>duplocloud_gcp_firestores</code> data <br>source.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-d53bac898b783067a8b4fca513969df0e0349ef096af752bbb237559c26ebc25">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource.tf</strong><dd><code>Terraform Configuration Example for GCP Firestore Resource</code></dd></summary>
<hr>
      
examples/resources/duplocloud_gcp_firestore/resource.tf

<li>Showcased a Terraform configuration for creating a GCP Firestore <br>resource.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/492/files#diff-516991d3110c36f72bfb4f9d4e386067f55472bf5c4d0c07c2edbd684ab2e1f3">+17/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-16561 [TF] GCP: Support for Firestore DB](https://app.clickup.com/t/8655600/DUPLO-16561)

File: docs/resources/gcp_firestore.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `location_id` (String) Location for firestore
+- `name` (String) The short name of the firestore.  Duplo will add a prefix to the name.  You can retrieve the full name from the `fullname` attribute.
+- `tenant_id` (String) The GUID of the tenant that the firestore will be created in.
+- `type` (String) Firestore type
```
Comment (komalkadam-dc) [inline docs/resources/gcp_firestore.md:43]: add possible values

**Review by komalkadam-dc — CHANGES_REQUESTED** [review]
Add possible values for type, rest looks good

File: docs/resources/gcp_firestore.md (hunk 1 above)
Comment (Duplo-nikv) [reply docs/resources/gcp_firestore.md]: done

**Review by komalkadam-dc — APPROVED** [review]

**Review by tahir-duplo — APPROVED** [review]
LGTM

**Pull request merged**